File: mobile/lib/models/book.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String year; // оставляем String, как приходит с бэкенда
  final String description;
  final String genre;
  final String? cover;
  final List<String>? photos;
  final int pages;

  Book({
    required this.id,
    required this.title,
    required this.author,
    required this.year,
    required this.description,
    required this.genre,
    this.cover,
    this.photos,
    required this.pages,
  });

  // Парсинг JSON в объект Dart
  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'] as String,
      title: json['title'] as String,
      author: json['author'] as String,
      year: json['year'] as String,
      description: json['description'] as String,
      genre: json['genre'] as String,
      cover: json['cover'] as String?,
      photos: json['photos'] != null ? List<String>.from(json['photos']) : null,
      pages: json['pages'] as int,
    );
  }

  // Преобразование объекта в JSON (если нужно для отправки на бэкенд)
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'author': author,
      'year': year,
      'description': description,
      'genre': genre,
      'cover': cover,
      'photos': photos,
      'pages': pages,
    };
  }
}

File: parser/main_package/models/book.py
import re
from typing import Optional, TypedDict


class Book(TypedDict):
    """
    универсальный формат книги — используется и для сырых, и для сгенерированных данных
    """
    id: Optional[str]
    title: str
    author: Optional[str]
    year: Optional[int]
    description: Optional[str]
    genre: Optional[list[str]]
    cover: Optional[str]
    pages: Optional[int]


def format_year(year: str | int) -> Optional[int]:
    """
    преобразует разные форматы годов (строки, диапазоны, BC и т.п.) в int
    """
    if not year:
        return None

    if isinstance(year, int):
        return year

    text = year.strip().replace("–", "-").lower()
    is_bc = any(x in text for x in ["bc", "до н"])  # проверка на "год до нашей эры"
    digits = [int(y) for y in re.findall(r"\d{1,4}", text)]  # получение всех чисел

    if not digits:
        return None

    year = digits[-1]  # если диапазон - берем последнее значение
    return -year if is_bc else year


def normalize_title(raw_book: Book) -> str:
    """
    упрощает название книги: убирает лишние слова и скобки
    """
    if not raw_book["title"]:
        return ""
    t = raw_book["title"].replace("Книга ", "")
    t = re.split(r"\s*\(|[:\-–—]\s*", t)[0]
    return t.strip()

